crates/rpl_patterns/src/cve_2020_25016.rs

```diff
@@ -93,57 +91,62 @@ impl<'tcx> Visitor<'tcx> for CheckFnCtxt<'_, 'tcx> {
     }
 }
 
-struct PatternCast {
+struct PatternCast<'pcx> {
+    pattern: MirPattern<'pcx>,
     ty_var: pat::TyVarIdx,
     cast_from: pat::Location,
     cast_to: pat::Location,
 }
 
-#[rpl_macros::mir_pattern]
-fn pattern_cast(patterns: &mut pat::MirPatternBuilder<'_>) -> PatternCast {
-    mir! {
-        meta!($T:ty);
-
-        let from_slice: &[$T] = _;
-        let from_raw: *const [$T] = &raw const *from_slice;
-        let from_len: usize = PtrMetadata(copy from_slice);
-        let ty_size: usize = SizeOf($T);
-        let to_ptr_t: *const T = move from_raw as *const $T (PtrToPtr);
-        let to_ptr: *const u8 = move to_ptr_t as *const u8 (PtrToPtr);
-        let to_len: usize = Mul(move from_len, move ty_size);
-        let to_raw: *const [u8] = *const [u8] from (copy to_ptr, copy to_len);
-        let to_slice: &[u8] = &*to_raw;
+#[rpl_macros::pattern_def]
+fn pattern_cast(pcx: PatCtxt<'_>) -> PatternCast<'_> {
+    rpl! {
+        fn $pattern (..) -> _ = mir! {
```

Comment:
@stuuupidcat @TheVeryDarkness Does this syntax make sense? (The function header is ignored now, and the `MirPattern` variable is given by the name `$pattern`)

Reply:
Intuitively/Formally, `$pattern`s position makes it more like the function's name than the pattern's name. Maybe we can temporarily use a seperate line (or other similar ways) to declare the pattern's name?

For example,

```rust 
fn pattern_cast(pcx: PatCtxt<'_>) -> PatternCast<'_> {
    rpl! {
        $pattern
        fn _ (..) -> _ = mir! {...}
    }
}
```

Reply:
Actually that should be
```
#[export(pattern)]
fn _ (..) -> _ = mir! { ... }
```
in the next version, and it still requires extra work to make `#[export(...)]` work.

It is just a temporary implementation for now.